fastlane_bot/pool_finder.py

```diff
@@ -0,0 +1,234 @@
+                                      each dictionary should align with the requirements of the _extract_pairs,
+                                      _find_unsupported_pairs, and _find_unsupported_triangles methods.
+
+        Returns:
+            Dict: Returns a dictionary with pools sorted into different exchange types (Uni V2 forks, Uni V3 forks,
+                  and Solidly V2 forks), each associated with their specific supporting pools based on the unsupported
+                  configurations identified.
+        """
+        carbon_pairs, other_pairs = self._extract_pairs(pools=pools, carbon_forks=self.carbon_forks)
+        if not carbon_pairs:
+            return [], [], []
+        self.extract_univ3_fee_tiers(pools)
+        func = PoolFinder._find_unsupported_triangles if arb_mode in ["triangle", "multi_triangle"] else PoolFinder._find_unsupported_pairs
+        unsupported = func(self.flashloan_tokens, carbon_pairs=carbon_pairs, external_pairs=other_pairs)
+        supporting_pools = self._find_pools(unsupported)
+        return self._sort_exchange_pools(supporting_pools, self.uni_v2_forks, self.uni_v3_forks, self.solidly_v2_forks)
+
+
+    def _find_pools(self, unsupported_pairs: List[Tuple]) -> Dict[str, List[str]]:
+        """
+            Collects pool addresses for each exchange, based on a set of unsupported token pairs
+            and flashloan tokens. The function constructs pairs of tokens from unsupported_pairs with each
+            flashloan token and retrieves pool data via multicall. It filters out invalid addresses.
+
+            Args:
+                unsupported_pairs (List[Tuple]): A list of tuples, where each tuple contains two token addresses.
+                flashloan_tokens (List[str]): A list of token addresses available for flashloans.
+
+            Returns:
+                Dict[str, List[str]]: A list of dictionaries, where each dictionary maps an exchange's name
+                                            to a list of valid pool addresses (i.e., non-zero addresses) obtained
+                                            from the multicall across all generated pairs.
+
+            Raises:
+                Exception: An exception could be raised from the multicall operation depending on the
+                           implementation specifics of the multicall context manager or the exchange's
+                           get_pool_function method if it encounters a problem.
+            """
+        pairs = [(tkn, token) for pair in unsupported_pairs for tkn in pair for token in self.flashloan_tokens]
+
+        result_list = {}
+        for ex_name, ex_data in self.multicallers.items():
+            mc = ex_data["multicaller"]
+            ex = ex_data["exchange"]
+            with mc:
+                for pair in pairs:
+                    if ex.base_exchange_name == UNISWAP_V2_NAME:
+                        mc.add_call(ex.get_pool_function(ex.sync_factory_contract), pair[0], pair[1])
```

Comment:
You may as well just use the contract function directly.

Reply:
Not relevant any more